src/Blocks/BlockLeaves.h

```diff
@@ -10,29 +10,85 @@
 // Leaves can be this many blocks that away (inclusive) from the log not to decay
 #define LEAVES_CHECK_DISTANCE 6
 
-#define PROCESS_NEIGHBOR(x, y, z) \
-	do { \
-		switch (a_Area.GetBlockType(x, y, z)) \
-		{ \
-			case E_BLOCK_LEAVES: a_Area.SetBlockType(x, y, z, static_cast<BLOCKTYPE>(E_BLOCK_SPONGE + i + 1)); break; \
-			case E_BLOCK_LOG: return true; \
-			case E_BLOCK_NEW_LEAVES: a_Area.SetBlockType(x, y, z, static_cast<BLOCKTYPE>(E_BLOCK_SPONGE + i + 1)); break; \
-			case E_BLOCK_NEW_LOG: return true; \
-		} \
-	} while (false)
 
-bool HasNearLog(cBlockArea &a_Area, int a_BlockX, int a_BlockY, int a_BlockZ);
 
 
 
-
-
-class cBlockLeavesHandler :
+class cBlockLeavesHandler:
 	public cBlockHandler
 {
+	using Super = cBlockHandler;
+
+	/** Returns true if the area contains a continous path from the specified block to a log block entirely made out of leaves blocks. */
+	static bool HasNearLog(cBlockArea & a_Area, const Vector3i a_BlockPos)
+	{
+		// Filter the blocks into a {leaves, log, other (air)} set:
+		auto * Types = a_Area.GetBlockTypes();
+		for (size_t i = a_Area.GetBlockCount() - 1; i > 0; i--)
+		{
+			switch (Types[i])
+			{
+				case E_BLOCK_LEAVES:
+				case E_BLOCK_LOG:
+				case E_BLOCK_NEW_LEAVES:
+				case E_BLOCK_NEW_LOG:
+				{
+					break;
+				}
+				default:
+				{
+					Types[i] = E_BLOCK_AIR;
+					break;
+				}
+			}
+		}  // for i - Types[]
+
+		// Perform a breadth-first search to see if there's a log connected within 4 blocks of the leaves block:
+		// Simply replace all reachable leaves blocks with a sponge block plus iteration (in the Area) and see if we can reach a log
+		a_Area.SetBlockType(a_BlockPos.x, a_BlockPos.y, a_BlockPos.z, E_BLOCK_SPONGE);
+		for (int i = 0; i < LEAVES_CHECK_DISTANCE; i++)
+		{
+			for (int y = std::max(a_BlockPos.y - i, 0); y <= std::min(a_BlockPos.y + i, cChunkDef::Height - 1); y++)
+			{
+				for (int z = a_BlockPos.z - i; z <= a_BlockPos.z + i; z++)
+				{
+					for (int x = a_BlockPos.x - i; x <= a_BlockPos.x + i; x++)
+					{
+						if (a_Area.GetBlockType(x, y, z) != E_BLOCK_SPONGE + i)
+						{
+							continue;
+						}
+						#define PROCESS_NEIGHBOR(x, y, z) \
```

Comment:
What was the rationale for this change? Surely the pre-processor doesn't respect scope

Reply:
I like to have the definition as close to the usage as possible, and `#undef` as soon as possible. Macros have no scoping, so they pollute the global namespace and thus are dangerous. Having the macro definition at the top of the file and the function at the bottom was needlessly far away.

Reply:
If you're doing that, why not just make it a lambda?

Reply:
The macro is doing `return` from the entire function, which doesn't translate too well to a lambda. But it's true that it looks better.